Remove 'IsIPv6Cluster' function

ormergi opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
We use `IsIPv6Cluster` in tests that use ISCSI PVC's and at probes tests
this function is no longer needed, this PR removes it.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

ormergi commented:
/cc @AlonaKaplan

oshoval commented on tests/storage_test.go:
```diff
@@ -26,6 +26,8 @@ import (
 	"time"
 
+	netutils "k8s.io/utils/net"
+
```
same

ormergi replied:
Done

qinqon commented on tests/probes_test.go:
```diff
@@ -51,17 +51,10 @@ var _ = Describe("[ref_id:1182]Probes", func() {
-			// 'connection reset by peer' when bound to IPv6 addresses. On those scenarios, we fallback to using a
-			// Fedora VM.
-			useFedoraVM := tests.IsIPv6Cluster(virtClient) && readinessProbe == httpProbe
+			// 'connection reset by peer' when bound to IPv6 addresses. Thus we use Fedora VM.
```
I think that we were safe using cirros and that the bug was not related to cirros but to something at content-length, @maiqueb knos better.

maiqueb replied:
I'll eventually provide a decent explanation on https://github.com/kubevirt/kubevirt/pull/4129/commits/b41e635714a240fc082c8b7228ca5e10c8817d3b .

oshoval replied:
dont we want to keep StartHTTPServer supporting the two VMs types ?
and if we dont, then the bool flag can be removed all over (i think i still saw it in couple places?)

maiqueb replied:
> dont we want to keep StartHTTPServer supporting the two VMs types ?

Do you have a reason to keep it ? I don't.

> and if we dont, then the bool flag can be removed all over (i think i still saw it in couple places?)
> https://github.com/kubevirt/kubevirt/pull/4151/files#diff-0836acab6741ba8251f478229fe81a6dR71

Have you seen the commit I pasted on https://github.com/kubevirt/kubevirt/pull/4151#discussion_r485528493 ? It removes that.

oshoval replied:
> > dont we want to keep StartHTTPServer supporting the two VMs types ?
> 
> Do you have a reason to keep it ? I don't.
> 
> > and if we dont, then the bool flag can be removed all over (i think i still saw it in couple places?)
> > https://github.com/kubevirt/kubevirt/pull/4151/files#diff-0836acab6741ba8251f478229fe81a6dR71
> 
> Have you seen the commit I pasted on [#4151 (comment)](https://github.com/kubevirt/kubevirt/pull/4151#discussion_r485528493) ? It removes that.

Ye saw, it confused me with the code on this PR,
maybe worth also to write which PR this commit belongs to (less confusing and we can see easily when its merged and so on)

oshoval commented on tests/migration_test.go:
```diff
@@ -692,6 +690,10 @@ var _ = Describe("[rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system
 				// Start a ISCSI POD and service
 				By("Starting an iSCSI POD")
 				iscsiIP := tests.CreateISCSITargetPOD(cd.ContainerDiskEmpty)
+				// Skip tests that use ISCSI PVC's with IPv6 until https://github.com/kubevirt/kubevirt/issues/4150 solved
```
Also please put all those occurrences in a function and use it

ormergi replied:
Done

ormergi commented on tests/migration_test.go:
```diff
@@ -26,6 +26,8 @@ import (
 	"strings"
 	"sync"
 
+	netutils "k8s.io/utils/net"
```
Done

ormergi commented on tests/migration_test.go:
```diff
@@ -809,13 +811,16 @@ var _ = Describe("[rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system
 				By("Starting an iSCSI POD")
 				iscsiIP := tests.CreateISCSITargetPOD(cd.ContainerDiskAlpine)
+
+				// Skip tests that use ISCSI PVC's with IPv6 until https://github.com/kubevirt/kubevirt/issues/4150 solved
```
Done

oshoval commented on tests/utils.go:
```diff
@@ -3316,6 +3316,13 @@ func RemoveHostDiskImage(diskPath string, nodeName string) {
 	Eventually(getStatus, 30, 1).Should(Equal(k8sv1.PodSucceeded))
 }
 
+// Skip tests that use ISCSI PVC's with IPv6 until https://github.com/kubevirt/kubevirt/issues/4150 solved
```
i would add this comment to the Skip
see please how we are doing this on all the other skippers

maiqueb replied:
+1; having the issue on the `Skip`s body is helpful.

ormergi replied:
Done

maiqueb commented on tests/probes_test.go:
```diff
@@ -75,7 +68,7 @@ var _ = Describe("[ref_id:1182]Probes", func() {
 
 			By("Starting the server inside the VMI")
-			serverStarter(vmi, 1500, useFedoraVM)
+			serverStarter(vmi, 1500, true)
```
I might be missing something here; how can you do this ? If there's only IPv4, you don't want to use a big needy fedora VM.

ormergi replied:
Since there is an issue with using cirros VM in this test on dual stack cluster then I think we should be consistent and use the image that work for both IP families instead of checking if a cluster is dual stack.

Eventually all providers will support dual stack and the issue with `netcat` on cirros will remain until we will fix it.
I think we should use only fedora for this test to keep it simple and not switching VM images because one wont work, 
we dont save much resources in this case it wont be noticeable.

maiqueb replied:
I haven't explained myself properly. There is no issue with netcat on fedora. You can just plainly use a cirros VM and it will work.

ormergi replied:
There is an issue using netcat with cirros image for this test, isnt it?

maiqueb replied:
It's explained better in https://github.com/kubevirt/kubevirt/pull/4129/commits/e956c1a282e2cf036fffd28a91d3977a147dd272

ormergi replied:
So we can use cirros for this test as well, I will change it

maiqueb reviewed (commented):
Mostly, what I don't understand is in https://github.com/kubevirt/kubevirt/pull/4151/commits/6e43c39ec0881dfd14e4e9475952e35d70a11b14 .

Instead of 'skipping' always, we now skip in each test. I'm aware you do this because you gotta check the presence of an IPv6 addr on the ISCSI pod.

I personally would rather keep the skips in the `BeforeEach` part, since it does the same with less code ... You could rebase this PR #4139 and use the `libnet.IsClusterDualStack` function to know if you're expected to skip or not.

If there's a test that works while other don't, then what I've just said makes no sense.

ormergi commented:
> Mostly, what I don't understand is in [6e43c39](https://github.com/kubevirt/kubevirt/commit/6e43c39ec0881dfd14e4e9475952e35d70a11b14) .
> 
> Instead of 'skipping' always, we now skip in each test. I'm aware you do this because you gotta check the presence of an IPv6 addr on the ISCSI pod.
> 
> I personally would rather keep the skips in the `BeforeEach` part, since it does the same with less code ... You could rebase this PR #4139 and use the `libnet.IsClusterDualStack` function to know if you're expected to skip or not.

Keeping `Skip` at before each will skip the test for dual stack clusters and we do want to run those tests on dual stack lane(s).
Since `CreateISCSITargetPod` returns `podIP` (at the current state) on dual stack and single ipv4 clusters it will always be IPv4 and the test will run, we should skip those tests only if `CreateISCSITargetPod` returns IPv6 address.

ormergi commented:
Changes:
- Rebased  
- Use cirros VM's in probe test (test id's 1200 and 1202)

oshoval commented:
Rebase please, lets see lanes pass

AlonaKaplan commented on tests/probes_test.go:
```diff
@@ -5,11 +5,11 @@ import (
 	. "github.com/onsi/gomega"
 	v1 "k8s.io/api/core/v1"
-	v13 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
How is this change related to the PR

ormergi replied:
At first I changes on of the tests in probes_tests.go so I fixed those libraries aliases 
v12 and v13 are auto generated names by the IDE

AlonaKaplan replied:
As far as I see, now the probles are not part of the PR anymore. The PR is dealing only with the ISCSI network related tests. Can you please remove this commit from the PR. It can be done in a separate PR.

AlonaKaplan replied:
Can you please remove this commit from the PR?

ormergi replied:
done

AlonaKaplan commented on tests/probes_test.go:
```diff
@@ -5,11 +5,11 @@ import (
 	netutils "k8s.io/utils/net"
 
-	v12 "kubevirt.io/client-go/api/v1"
+	kvapiv1 "kubevirt.io/client-go/api/v1"
```
same

AlonaKaplan commented on tests/migration_test.go:
```diff
@@ -692,6 +686,8 @@ var _ = Describe("[rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system
 				// Start a ISCSI POD and service
 				By("Starting an iSCSI POD")
 				iscsiIP := tests.CreateISCSITargetPOD(cd.ContainerDiskEmpty)
+				tests.SkipIscsiPVCWithIPv6Test(iscsiIP)
```
As we are going to drop the ipv6 only cluster this skip is redundant.
My initial thought was to prepare the test to be dual stack (work with both of the pod ips) but skip just the ipv6 one. But let's make it even easier, change the description of the relevant tests to mention it tests only the ipv4 flow.
For example here - 
`Context("with a shared ISCSI Filesystem PVC using the ipv4 address of the ISCSI pod...`

ormergi replied:
We are on the same page, I tried to make those test work for both IP families but I there are issues when using IPv6 for ISCSI PVCs (#4150 ), so until we figure out how to fix them I added a skip

I will update the context description in following commit

oshoval replied:
if you remove the skip totally, wont it fail on kind ipv6 until the lane is removed ?
(unless you plan to hold this one or remove the kind in the same PR which isnt good imo)

ormergi replied:
We dont want to remove this skip until single stack IPv6 lane is dropped

oshoval replied:
"As we are going to drop the ipv6 only cluster this skip is redundant."
so this, its better than master have (the skip there is less accurate)
so i am confused a bit

AlonaKaplan replied:
This test is a ipv4 only test.  For me it doesn't make sense to skip it in case the podIp is ipv6 one. The test should always test the ipv4 address of the pod. If the pod doesn't have ipv4 adrees it should be skipped. But since we don't have such a cluster with no ipv4 address on the pod we can assume the pod has ipv4 address and fail otherwise.

I would use [ipAddressess.GetPodIpByFamily](https://github.com/kubevirt/kubevirt/blob/6a8d40ed2956bfd5b3d8e75a4925e91fd6866904/tests/libnet/ipaddress.go#L8) to get the ipv4 address of the pod (you can rebase on to of the mentioned unmerged PR).

Another option is to fail the test if the podIp is not an ipv4 one. I like this option less because we will have to change the test one a cluster with ipv6 as the primary ip will be introduced.
But since we currently support only clusters with ipv4 as the primary one, this option is valid.

oshoval replied:
i would just keep the PR as its now, because it will skip if the ip under testing is ipv6, and it can happen miss only dual stack with reverse order that doesnt exists.
we already have tests that have the same problem as well, they would just fail in case we would have a reverse cluster, but i dont see it happen in the near future, and until it will , the needed PRs will be merged and it will be easy to fix what needed.

this way we wont block, and not create dependency
wdyt?

AlonaKaplan replied:
It doesn't make sense to keep this skip here. The test is ipv4 only, if is confusing to skip the test if the address is ipv6. Why would the address be ipv6 on an ipv4 test. 
Just to illustrate the problem. Once the isci ipv6 bug will be fixed, can you just remove the skip? Would this ipv4 test run with ipv4 address.
As I already wrote, since the test is ipv4 one, you need to make sure you [use the ipv4 address](https://github.com/kubevirt/kubevirt/blob/6a8d40ed2956bfd5b3d8e75a4925e91fd6866904/tests/libnet/ipaddress.go#L8).

ormergi replied:
I dont expect it to work by just removing the skip when the IPv6 issue will be fixed, we cant know what changes there would be once it is resolved (e.g: API change).

Since this PR scope is removing IsIPv6Cluster as prepration for removing ipv6 only lane and ISCSI PV wont work with IPv6 for now, we should not waste time in adapting or changing ISCSI PV tests to support dual-stack.

We can remove `SkipIscsiPVCWithIPv6Test` function and use `SkipOnKindInfra` instead 
Then we will end up with those tests working with IPv4 address only on dual-stack and single ipv4 stack cluster
with explicit title.
What you think?

AlonaKaplan replied:
> I dont expect it to work by just removing the skip when the IPv6 issue will be fixed, we cant know what changes there would be once it is resolved (e.g: API change).

We do know the test is ipv4 only so skipping it when the ip is ipv6 because ipv6 has a bug doesn't make any sense.
This test should never be invoked for ipv6 since it is ipv4 test. It has nothing to do with the bug.

> 
> Since this PR scope is removing IsIPv6Cluster as preparation for removing ipv6 only lane and ISCSI PV wont work with IPv6 for now, we should not waste time in adapting or changing ISCSI PV tests to support dual-stack.
> 

The lane was already removed.
I didn't ask you to adapt the test to support dual stack. The opposite is correct. I asked this test to be ipv4 only.
You can find [here](https://github.com/kubevirt/kubevirt/pull/4156/files#diff-d642ef9101fdb0671d818d73a1c3f977R991) an example of how we get the ip address from a pod.

I will summarize again the options -
1. Using the PodIp
2. Using [libnet.GetPodIpByFamily(pod, k8sv1.IPv4Protocol) ](https://github.com/kubevirt/kubevirt/blob/6a8d40ed2956bfd5b3d8e75a4925e91fd6866904/tests/libnet/ipaddress.go#L8) to get the ipv4 address of the pod.

I prefer the second option since we know that in the future the main pod ip address may be the ipv6 one and it is pretty cheap  not to reside on the fact currently the ipv4 address is the primary one.
I would also like all our ipv4 only tests to be aligned an use the same pattern.

> We can remove `SkipIscsiPVCWithIPv6Test` function and use `SkipOnKindInfra` instead
> Then we will end up with those tests working with IPv4 address only on dual-stack and single ipv4 stack cluster
> with explicit title.
> What you think?

I think the is no reason to do it since the ipv6 kind lane is already gone.

ormergi replied:
Gotcha

maiqueb commented:
> > Mostly, what I don't understand is in [6e43c39](https://github.com/kubevirt/kubevirt/commit/6e43c39ec0881dfd14e4e9475952e35d70a11b14) .
> > Instead of 'skipping' always, we now skip in each test. I'm aware you do this because you gotta check the presence of an IPv6 addr on the ISCSI pod.
> > I personally would rather keep the skips in the `BeforeEach` part, since it does the same with less code ... You could rebase this PR #4139 and use the `libnet.IsClusterDualStack` function to know if you're expected to skip or not.
> 
> Keeping `Skip` at before each will skip the test for dual stack clusters and we do want to run those tests on dual stack lane(s).
> Since `CreateISCSITargetPod` returns `podIP` (at the current state) on dual stack and single ipv4 clusters it will always be IPv4 and the test will run, we should skip those tests only if `CreateISCSITargetPod` returns IPv6 address.

But you don't need to check that from the `CreateISCSITargetPod` pod - that exact same behavior will happen if you check the `podIP` on a virt-handler. Personally, I'd prefer if you'd do something like https://github.com/kubevirt/kubevirt/blob/master/tests/libnet/cluster.go#L15 but only getting the podIP (or assuring you having a single IP).

ormergi commented:
> CreateISCSITargetPod



> > > Mostly, what I don't understand is in [6e43c39](https://github.com/kubevirt/kubevirt/commit/6e43c39ec0881dfd14e4e9475952e35d70a11b14) .
> > > Instead of 'skipping' always, we now skip in each test. I'm aware you do this because you gotta check the presence of an IPv6 addr on the ISCSI pod.
> > > I personally would rather keep the skips in the `BeforeEach` part, since it does the same with less code ... You could rebase this PR #4139 and use the `libnet.IsClusterDualStack` function to know if you're expected to skip or not.
> > 
> > 
> > Keeping `Skip` at before each will skip the test for dual stack clusters and we do want to run those tests on dual stack lane(s).
> > Since `CreateISCSITargetPod` returns `podIP` (at the current state) on dual stack and single ipv4 clusters it will always be IPv4 and the test will run, we should skip those tests only if `CreateISCSITargetPod` returns IPv6 address.
> 
> But you don't need to check that from the `CreateISCSITargetPod` pod - that exact same behavior will happen if you check the `podIP` on a virt-handler. Personally, I'd prefer if you'd do something like https://github.com/kubevirt/kubevirt/blob/master/tests/libnet/cluster.go#L15 but only getting the podIP (or assuring you having a single IP).

So what you suggest is to check if the cluster is dual-stack using the function you mentioned before the tests starts?
BTW this is how we skip those test before using `isIPv6Cluster`

maiqueb commented:
> > CreateISCSITargetPod
> 
> > > > Mostly, what I don't understand is in [6e43c39](https://github.com/kubevirt/kubevirt/commit/6e43c39ec0881dfd14e4e9475952e35d70a11b14) .
> > > > Instead of 'skipping' always, we now skip in each test. I'm aware you do this because you gotta check the presence of an IPv6 addr on the ISCSI pod.
> > > > I personally would rather keep the skips in the `BeforeEach` part, since it does the same with less code ... You could rebase this PR #4139 and use the `libnet.IsClusterDualStack` function to know if you're expected to skip or not.
> > > 
> > > 
> > > Keeping `Skip` at before each will skip the test for dual stack clusters and we do want to run those tests on dual stack lane(s).
> > > Since `CreateISCSITargetPod` returns `podIP` (at the current state) on dual stack and single ipv4 clusters it will always be IPv4 and the test will run, we should skip those tests only if `CreateISCSITargetPod` returns IPv6 address.
> > 
> > 
> > But you don't need to check that from the `CreateISCSITargetPod` pod - that exact same behavior will happen if you check the `podIP` on a virt-handler. Personally, I'd prefer if you'd do something like https://github.com/kubevirt/kubevirt/blob/master/tests/libnet/cluster.go#L15 but only getting the podIP (or assuring you having a single IP).
> 
> So what you suggest is to check if the cluster is dual-stack using the function you mentioned before the tests starts?
> BTW this is how we skip those test before using `isIPv6Cluster`



> > CreateISCSITargetPod
> 
> > > > Mostly, what I don't understand is in [6e43c39](https://github.com/kubevirt/kubevirt/commit/6e43c39ec0881dfd14e4e9475952e35d70a11b14) .
> > > > Instead of 'skipping' always, we now skip in each test. I'm aware you do this because you gotta check the presence of an IPv6 addr on the ISCSI pod.
> > > > I personally would rather keep the skips in the `BeforeEach` part, since it does the same with less code ... You could rebase this PR #4139 and use the `libnet.IsClusterDualStack` function to know if you're expected to skip or not.
> > > 
> > > 
> > > Keeping `Skip` at before each will skip the test for dual stack clusters and we do want to run those tests on dual stack lane(s).
> > > Since `CreateISCSITargetPod` returns `podIP` (at the current state) on dual stack and single ipv4 clusters it will always be IPv4 and the test will run, we should skip those tests only if `CreateISCSITargetPod` returns IPv6 address.
> > 
> > 
> > But you don't need to check that from the `CreateISCSITargetPod` pod - that exact same behavior will happen if you check the `podIP` on a virt-handler. Personally, I'd prefer if you'd do something like https://github.com/kubevirt/kubevirt/blob/master/tests/libnet/cluster.go#L15 but only getting the podIP (or assuring you having a single IP).
> 
> So what you suggest is to check if the cluster is dual-stack using the function you mentioned before the tests starts?
> BTW this is how we skip those test before using `isIPv6Cluster`

I may have lost the scope of this PR, please keep me honest; you're writing a function that checks if the cluster is deploying using pure IPv6 configuration (no IPv4) - that doesn't check the dns service (what we currently have); it checks instead the network configuration of a pod you've created.

I'm saying that you don't need to check that pod - any pod on the system will give you the same information: virt-handler for instance. 

And no, I don't want you to use that function; I want you to follow those principles, and write one suited to your scenario - which is to skip on 'pure' IPv6. 

Did I somehow miss-understand what you're attempting to do?

ormergi commented:
> 
> I may have lost the scope of this PR, please keep me honest; you're writing a function that checks if the cluster is deploying using pure IPv6 configuration (no IPv4) - that doesn't check the dns service (what we currently have); it checks instead the network configuration of a pod you've created.
> 
> I'm saying that you don't need to check that pod - any pod on the system will give you the same information: virt-handler for instance.
> 
> And no, I don't want you to use that function; I want you to follow those principles, and write one suited to your scenario - which is to skip on 'pure' IPv6.
> 
> Did I somehow miss-understand what you're attempting to do?

This PR scope is `isIPv6Cluster` function as preparation for removing `kind-k8s-1.17-ipv6` lane.

The function  `SkipIscsiPVCWithIPv6Test` is not about checking if the cluster is dual stack, I was asked to extract that check to a function.

I do follow the principals you mentioned by creating such function which is `SkipIscsiPVCWithIPv6Test`
the other way to skip those tests is by using `IsDualStackCluster` which you dont want.

My scenario is not `skip if pure IPv6`, its about skipping tests that use ISCSI PV configured with IPv6.
In all tests that use ISCSI PV's, the pattern is creating iscsi target pod, based on that pod IP address 
create PV and PVC.
In case that 'CreateISCSITargetPod' returned IPv6 we need to skip those tests (#4150)
I do not adapt tests that use `CreateISCSITargetPod` to dual stack because it wont work for the IPv6 part 
this is why skipping if the iscsi target pod is IPv6  may look awkward to you

maiqueb commented:
> > I may have lost the scope of this PR, please keep me honest; you're writing a function that checks if the cluster is deploying using pure IPv6 configuration (no IPv4) - that doesn't check the dns service (what we currently have); it checks instead the network configuration of a pod you've created.
> > I'm saying that you don't need to check that pod - any pod on the system will give you the same information: virt-handler for instance.
> > And no, I don't want you to use that function; I want you to follow those principles, and write one suited to your scenario - which is to skip on 'pure' IPv6.
> > Did I somehow miss-understand what you're attempting to do?
> 
> This PR scope is `isIPv6Cluster` function as preparation for removing `kind-k8s-1.17-ipv6` lane.
> 
> The function `SkipIscsiPVCWithIPv6Test` is not about checking if the cluster is dual stack, I was asked to extract that check to a function.
> 
> I do follow the principals you mentioned by creating such function which is `SkipIscsiPVCWithIPv6Test`
> the other way to skip those tests is by using `IsDualStackCluster` which you dont want.
> 
> My scenario is not `skip if pure IPv6`, its about skipping tests that use ISCSI PV configured with IPv6.
> In all tests that use ISCSI PV's, the pattern is creating iscsi target pod, based on that pod IP address
> create PV and PVC.

My point is that if virt-handler has a single ipv6 address, then the pod you're creating for the ISCSI pod **will also** have a single IPv6 addres. Thus, you can perform the skip on a single place, instead of having to create the pod, and only afterwards skip the test.

Other than requiring less code, it reaches the same conclusion faster - your proposed solution does meaningless work on an ipv6 cluster (ok, we want to remove that lane, but still .... ) since it creates a pod, **then** skips the test.

> In case that 'CreateISCSITargetPod' returned IPv6 we need to skip those tests (#4150)
> I do not adapt tests that use `CreateISCSITargetPod` to dual stack because it wont work for the IPv6 part
> this is why skipping if the iscsi target pod is IPv6 may look awkward to you

ormergi commented:
> > > I may have lost the scope of this PR, please keep me honest; you're writing a function that checks if the cluster is deploying using pure IPv6 configuration (no IPv4) - that doesn't check the dns service (what we currently have); it checks instead the network configuration of a pod you've created.
> > > I'm saying that you don't need to check that pod - any pod on the system will give you the same information: virt-handler for instance.
> > > And no, I don't want you to use that function; I want you to follow those principles, and write one suited to your scenario - which is to skip on 'pure' IPv6.
> > > Did I somehow miss-understand what you're attempting to do?
> > 
> > 
> > This PR scope is `isIPv6Cluster` function as preparation for removing `kind-k8s-1.17-ipv6` lane.
> > The function `SkipIscsiPVCWithIPv6Test` is not about checking if the cluster is dual stack, I was asked to extract that check to a function.
> > I do follow the principals you mentioned by creating such function which is `SkipIscsiPVCWithIPv6Test`
> > the other way to skip those tests is by using `IsDualStackCluster` which you dont want.
> > My scenario is not `skip if pure IPv6`, its about skipping tests that use ISCSI PV configured with IPv6.
> > In all tests that use ISCSI PV's, the pattern is creating iscsi target pod, based on that pod IP address
> > create PV and PVC.
> 
> My point is that if virt-handler has a single ipv6 address, then the pod you're creating for the ISCSI pod **will also** have a single IPv6 addres. Thus, you can perform the skip on a single place, instead of having to create the pod, and only afterwards skip the test.
> 
> Other than requiring less code, it reaches the same conclusion faster - your proposed solution does meaningless work on an ipv6 cluster (ok, we want to remove that lane, but still .... ) since it creates a pod, **then** skips the test.
> 
> > In case that 'CreateISCSITargetPod' returned IPv6 we need to skip those tests (#4150)
> > I do not adapt tests that use `CreateISCSITargetPod` to dual stack because it wont work for the IPv6 part
> > this is why skipping if the iscsi target pod is IPv6 may look awkward to you

Got it, I can drop that skip and the test will work for both single stack IPv6 and dual stack, 
by using IPv4 for those PV tests but thats fine now that the context name changed we have some kind of documentation.
How is that?

(@AlonaKaplan @oshoval  I guess this is what you meant in other discussion)

oshoval commented:
> > > > I may have lost the scope of this PR, please keep me honest; you're writing a function that checks if the cluster is deploying using pure IPv6 configuration (no IPv4) - that doesn't check the dns service (what we currently have); it checks instead the network configuration of a pod you've created.
> > > > I'm saying that you don't need to check that pod - any pod on the system will give you the same information: virt-handler for instance.
> > > > And no, I don't want you to use that function; I want you to follow those principles, and write one suited to your scenario - which is to skip on 'pure' IPv6.
> > > > Did I somehow miss-understand what you're attempting to do?
> > > 
> > > 
> > > This PR scope is `isIPv6Cluster` function as preparation for removing `kind-k8s-1.17-ipv6` lane.
> > > The function `SkipIscsiPVCWithIPv6Test` is not about checking if the cluster is dual stack, I was asked to extract that check to a function.
> > > I do follow the principals you mentioned by creating such function which is `SkipIscsiPVCWithIPv6Test`
> > > the other way to skip those tests is by using `IsDualStackCluster` which you dont want.
> > > My scenario is not `skip if pure IPv6`, its about skipping tests that use ISCSI PV configured with IPv6.
> > > In all tests that use ISCSI PV's, the pattern is creating iscsi target pod, based on that pod IP address
> > > create PV and PVC.
> > 
> > 
> > My point is that if virt-handler has a single ipv6 address, then the pod you're creating for the ISCSI pod **will also** have a single IPv6 addres. Thus, you can perform the skip on a single place, instead of having to create the pod, and only afterwards skip the test.
> > Other than requiring less code, it reaches the same conclusion faster - your proposed solution does meaningless work on an ipv6 cluster (ok, we want to remove that lane, but still .... ) since it creates a pod, **then** skips the test.
> > > In case that 'CreateISCSITargetPod' returned IPv6 we need to skip those tests (#4150)
> > > I do not adapt tests that use `CreateISCSITargetPod` to dual stack because it wont work for the IPv6 part
> > > this is why skipping if the iscsi target pod is IPv6 may look awkward to you
> 
> Got it, I can drop that skip and the test will work for both single stack IPv6 and dual stack,
> by using IPv4 for those PV tests but thats fine now that the context name changed we have some kind of documentation.
> How is that?
> 
> (@AlonaKaplan @oshoval I guess this is what you meant in other discussion)

but on kind you dont have ipv4 so how would it work?

ormergi commented:
> > > > > I may have lost the scope of this PR, please keep me honest; you're writing a function that checks if the cluster is deploying using pure IPv6 configuration (no IPv4) - that doesn't check the dns service (what we currently have); it checks instead the network configuration of a pod you've created.
> > > > > I'm saying that you don't need to check that pod - any pod on the system will give you the same information: virt-handler for instance.
> > > > > And no, I don't want you to use that function; I want you to follow those principles, and write one suited to your scenario - which is to skip on 'pure' IPv6.
> > > > > Did I somehow miss-understand what you're attempting to do?
> > > > 
> > > > 
> > > > This PR scope is `isIPv6Cluster` function as preparation for removing `kind-k8s-1.17-ipv6` lane.
> > > > The function `SkipIscsiPVCWithIPv6Test` is not about checking if the cluster is dual stack, I was asked to extract that check to a function.
> > > > I do follow the principals you mentioned by creating such function which is `SkipIscsiPVCWithIPv6Test`
> > > > the other way to skip those tests is by using `IsDualStackCluster` which you dont want.
> > > > My scenario is not `skip if pure IPv6`, its about skipping tests that use ISCSI PV configured with IPv6.
> > > > In all tests that use ISCSI PV's, the pattern is creating iscsi target pod, based on that pod IP address
> > > > create PV and PVC.
> > > 
> > > 
> > > My point is that if virt-handler has a single ipv6 address, then the pod you're creating for the ISCSI pod **will also** have a single IPv6 addres. Thus, you can perform the skip on a single place, instead of having to create the pod, and only afterwards skip the test.
> > > Other than requiring less code, it reaches the same conclusion faster - your proposed solution does meaningless work on an ipv6 cluster (ok, we want to remove that lane, but still .... ) since it creates a pod, **then** skips the test.
> > > > In case that 'CreateISCSITargetPod' returned IPv6 we need to skip those tests (#4150)
> > > > I do not adapt tests that use `CreateISCSITargetPod` to dual stack because it wont work for the IPv6 part
> > > > this is why skipping if the iscsi target pod is IPv6 may look awkward to you
> > 
> > 
> > Got it, I can drop that skip and the test will work for both single stack IPv6 and dual stack,
> > by using IPv4 for those PV tests but thats fine now that the context name changed we have some kind of documentation.
> > How is that?
> > (@AlonaKaplan @oshoval I guess this is what you meant in other discussion)
> 
> but on kind you dont have ipv4 so how would it work?

It wont this is why we need the skip for now :man_shrugging:

oshoval commented:
> > > > > > I may have lost the scope of this PR, please keep me honest; you're writing a function that checks if the cluster is deploying using pure IPv6 configuration (no IPv4) - that doesn't check the dns service (what we currently have); it checks instead the network configuration of a pod you've created.
> > > > > > I'm saying that you don't need to check that pod - any pod on the system will give you the same information: virt-handler for instance.
> > > > > > And no, I don't want you to use that function; I want you to follow those principles, and write one suited to your scenario - which is to skip on 'pure' IPv6.
> > > > > > Did I somehow miss-understand what you're attempting to do?
> > > > > 
> > > > > 
> > > > > This PR scope is `isIPv6Cluster` function as preparation for removing `kind-k8s-1.17-ipv6` lane.
> > > > > The function `SkipIscsiPVCWithIPv6Test` is not about checking if the cluster is dual stack, I was asked to extract that check to a function.
> > > > > I do follow the principals you mentioned by creating such function which is `SkipIscsiPVCWithIPv6Test`
> > > > > the other way to skip those tests is by using `IsDualStackCluster` which you dont want.
> > > > > My scenario is not `skip if pure IPv6`, its about skipping tests that use ISCSI PV configured with IPv6.
> > > > > In all tests that use ISCSI PV's, the pattern is creating iscsi target pod, based on that pod IP address
> > > > > create PV and PVC.
> > > > 
> > > > 
> > > > My point is that if virt-handler has a single ipv6 address, then the pod you're creating for the ISCSI pod **will also** have a single IPv6 addres. Thus, you can perform the skip on a single place, instead of having to create the pod, and only afterwards skip the test.
> > > > Other than requiring less code, it reaches the same conclusion faster - your proposed solution does meaningless work on an ipv6 cluster (ok, we want to remove that lane, but still .... ) since it creates a pod, **then** skips the test.
> > > > > In case that 'CreateISCSITargetPod' returned IPv6 we need to skip those tests (#4150)
> > > > > I do not adapt tests that use `CreateISCSITargetPod` to dual stack because it wont work for the IPv6 part
> > > > > this is why skipping if the iscsi target pod is IPv6 may look awkward to you
> > > 
> > > 
> > > Got it, I can drop that skip and the test will work for both single stack IPv6 and dual stack,
> > > by using IPv4 for those PV tests but thats fine now that the context name changed we have some kind of documentation.
> > > How is that?
> > > (@AlonaKaplan @oshoval I guess this is what you meant in other discussion)
> > 
> > 
> > but on kind you dont have ipv4 so how would it work?
> 
> It wont this is why we need the skip for now

of course but again i saw "Got it, I can drop that skip" so it raised an alert.

AlonaKaplan commented on tests/migration_test.go:
```diff
@@ -692,6 +686,8 @@ var _ = Describe("[rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system
 				// Start a ISCSI POD and service
 				By("Starting an iSCSI POD")
 				iscsiIP := tests.CreateISCSITargetPOD(cd.ContainerDiskEmpty)
+				tests.SkipIscsiPVCWithIPv6Test(iscsiIP)
```
Let's wait with this PR till 
https://github.com/kubevirt/project-infra/pull/642
https://github.com/kubevirt/kubevirt/pull/4156
are merged.
The we can assume there is not single stack ipv6 lane. And we can use [libnet.GetIp](https://github.com/kubevirt/kubevirt/pull/4156/files#diff-e96bc4003adf459cfc269a13a00c944bR8) to get the ipv4 address of the pod. I prefer it over assuming the podip is ipv4 one.

ormergi replied:
I dont think we should wait for #4156, until #4150 is resolved there is no point in changing `CreateISCSITargetPod`.
Why assuming that podIP is ipv4 is wrong?

oshoval replied:
its only a limitation atm, 
https://github.com/containerd/cri/issues/1278
at the future the primary might be ipv6

oshoval replied:
imo if you dont want to wait you can dual the tests, and add a skip on ipv6 part
this way you wont be depended on nothing, 
but you shouldnt assume the order please, and you need to use DescribeTable as its the new method we should use unless we decide otherwise.

ormergi replied:
I would agree with you if this test would work using IPv6 address for that ISCSI PV, but it doesn't so I dont think we should waste time adapting a test that is not going to work.

gabrielecerami commented:
@ormergi: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

gabrielecerami commented:
@ormergi: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

AlonaKaplan commented on tests/utils.go:
```diff
@@ -4637,11 +4637,6 @@ func RandTmpDir() string {
 	return tmpPath + "/" + rand.String(10)
 }
 
-func IsIPv6Cluster(virtClient kubecli.KubevirtClient) bool {
```
It makes sense to add the commit removing the method after there is no more code using this method:) But not critical to me as we are working in a PR scope.

ormergi replied:
Thanks for explaining :)

ormergi commented:
Rebsed
@AlonaKaplan

AlonaKaplan commented:
@ormergi please notice https://github.com/kubevirt/kubevirt/pull/4156 was merged

ormergi commented:
Rebase only

ormergi commented:
Changes:
- Removed skips from ISCSI PV tests that checks if the cluster is IPv6 and related commits
- Align with `CreateNFSTargetPod`: 
     - Changed `CreateISCSITargetPod` to return pod objects instead of IP address
     - Make use of `libnet.GetPodIpByFamily`
- Changed tests title to specify using ISCSI target pod IPv4 address

alonSadan commented:
/retest

AlonaKaplan commented on tests/migration_test.go:
```diff
@@ -697,18 +697,14 @@ var _ = Describe("[Serial][rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][leve
 			})
 		})
-		Context("with a shared ISCSI Filesystem PVC", func() {
+		Context("with a shared ISCSI Filesystem PVC (using ISCSI IPv4 address)", func() {
```
Didn't find a better place to comment on the commit message. There is no single stack ipv6 lane. We don't support it anymore. We remove the skip because we changed the test to be ipv4 only not because now it will work on single stack ip lane.

ormergi replied:
Done

AlonaKaplan commented:
/approve
/hold please just fix the small comment regarding one of the commit messages. Once done, you can remove the hold.

ormergi commented:
> /approve
> /hold please just fix the small comment regarding one of the commit messages. Once done, you can remove the hold.

Thanks 
/unhold

alonSadan commented on tests/migration_test.go:
```diff
@@ -725,8 +721,11 @@ var _ = Describe("[Serial][rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][leve
 				By("Starting an iSCSI POD")
-				iscsiIP := tests.CreateISCSITargetPOD(cd.ContainerDiskEmpty)
-				_, err = virtClient.CoreV1().PersistentVolumes().Create(tests.NewISCSIPV(pvName, "2Gi", iscsiIP, k8sv1.ReadWriteMany, k8sv1.PersistentVolumeFilesystem))
+				iscsiTargetPod := tests.CreateISCSITargetPOD(cd.ContainerDiskEmpty)
```
Why you changed to `cd.ContainerDiskEmpty` from  `cd.ContainerDiskCirros`? wouldn't that effect the ENV of the container in the ISCSI pod?

ormergi replied:
Its `ContainerDiskEmpty` on master https://github.com/kubevirt/kubevirt/blob/master/tests/migration_test.go#L728

alonSadan replied:
Right, but that's the only place, Isaw you changed it in some more places to be `cd.ContainerDiskEmpty` the take a look here for example : 
https://github.com/kubevirt/kubevirt/blob/master/tests/migration_test.go#L1364

ormergi replied:
Sorry this is my bad I will fix it, thanks!

alonSadan commented on tests/migration_test.go:
```diff
@@ -725,8 +721,11 @@ var _ = Describe("[Serial][rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][leve
-				iscsiIP := tests.CreateISCSITargetPOD(cd.ContainerDiskEmpty)
-				_, err = virtClient.CoreV1().PersistentVolumes().Create(tests.NewISCSIPV(pvName, "2Gi", iscsiIP, k8sv1.ReadWriteMany, k8sv1.PersistentVolumeFilesystem))
+				iscsiTargetPod := tests.CreateISCSITargetPOD(cd.ContainerDiskEmpty)
+				iscsiTargetIPAddress := libnet.GetPodIpByFamily(iscsiTargetPod, k8sv1.IPv4Protocol)
```
nit: 
could be in one line: 
`Expect(libnet.GetPodIpByFamily(iscsiTargetPod, k8sv1.IPv4Protocol)).NotTo(BeEmpty())`

ormergi replied:
It is necessary to pass `iscsiTargetIPAddress` to `NewISCSIPV` (line 728)

alonSadan replied:
Right, sorry,  my bad

alonSadan reviewed (commented):
Thanks OR. Looks good!
I added one question and one nit .

alonSadan reviewed (commented):
Thanks for making the changes.
/lgtm
Now let's wait for greens.

ormergi commented:
> Thanks for making the changes.
> /lgtm
> Now let's wait for greens.

Thanks :)
